odl/solvers/advanced/proximal_operators.py

```diff
@@ -369,15 +730,57 @@ def _call(self, x, out):
                 out.ufunc.maximum(lam, out=out)
 
                 # Global scaling
-                out /= lam
+                out /= (lam)
 
                 # Pointwise division
                 out.divide(diff, out)
 
     return _ProximalConvConjL1
 
 
-# TODO: move notes to ODL doc
+def proximal_l1(space, lam=1, g=None):
+    """Proximal operator factory of the l1-semi-norm.
+
+    Function for the proximal operator of the functional F where F is an
+    l1-semi-norm
+
+        F(x) = lam || ||x-g||_p ||_1
+
+    with x and g elements in ``space``, scaling factor lam, and point-wise
+    magnitude ||x||_p of x. If x is vector-valued, ||x||_p is the point-wise
```

Comment:
I'd say as long as it's an implementation detail, it can be omitted in
the description. If users can choose it, then it needs a decent
description. Our shipped guy (which could be used btw) is called
`PointwiseNorm`, so a natural parameter name would be `pointwise_norm`
or short `pw_norm`. Perhaps simply expect an operator as parameter?

On 2016-05-21 22:55, Jonas Adler wrote:
> In odl/solvers/advanced/proximal_operators.py
> <https://github.com/odlgroup/odl/pull/401#discussion_r64140164>:
> 
>>  
>>                  # Pointwise division
>>                  out.divide(diff, out)
>>  
>>      return _ProximalConvConjL1
>>  
>>  
>> -# TODO: move notes to ODL doc
>> +def proximal_l1(space, lam=1, g=None):
>> +    """Proximal operator factory of the l1-semi-norm.
>> +
>> +    Function for the proximal operator of the functional F where F is an
>> +    l1-semi-norm
>> +
>> +        F(x) = lam || ||x-g||_p ||_1
>> +
>> +    with x and g elements in ``space``, scaling factor lam, and point-wise
>> +    magnitude ||x||_p of x. If x is vector-valued, ||x||_p is the point-wise
> 
> Because this code implicitly does reductions for productspaces, This
> should be changed to be chosable, perhaps with its own parameter
> (instead of doing this implicitly depending on if |space| is a
> |ProductSpace|.
> 
> Suggestions for name of parameter:
> 
>   * reduce
>   * inner_norm
>   * more?
> 
> I wont touch this in the meanwhile.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/odlgroup/odl/pull/401/files/5a721d1d1325be17940d8eeb33e35bdd4879ad8a#r64140164>
> 

-- 

Dr Holger Kohr
KTH - Royal Institute of Technology
Dpt of Mathematics
100 44 Stockholm
Tel +46 8 790-6652
Mobil +46 72 0119332